Edge cases with 0-vectors in knn_search

While playing around with the new vector search endpoint I ran into two interesting edge cases with "cosine" similarity. 
I myabe didn't read the docs too carefully so I used zero-valued vectors in the query which raised an assertion error on my locally started nodes
where we run with "-ea" turned on. If assertions are disabled, and start with this setup

<details>

```
DELETE my-index

PUT my-index
{
  "mappings": {
    "properties": {
      "my_vector": {
        "type": "dense_vector",
        "dims": 3,
        "index": true,
        "similarity": "cosine"
      },
      "my_text" : {
        "type" : "keyword"
      }
    }
  }
}

PUT my-index/_doc/1
{
  "my_text" : "text1",
  "my_vector" : [3, 0, 6]
}

PUT my-index/_doc/2
{
  "my_text" : "text2",
  "my_vector" : [0, 1, 0]
}


POST my-index/_knn_search
{
  "knn": {
    "field": "my_vector",
    "query_vector": [0, 0, 0],
    "k": 2,
    "num_candidates" : 100
  }
}
```
</details>

I get <details><summary>this error</summary>

```
org.elasticsearch.action.search.SearchPhaseExecutionException: all shards failed
        [...]
Caused by: org.elasticsearch.ElasticsearchException$1: Index 2147483645 out of bounds for length 1
        at org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:639) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
        at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:410) [elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
        ... 20 more
Caused by: java.lang.IndexOutOfBoundsException: Index 2147483645 out of bounds for length 1
        at jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64) ~[?:?]
        at jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70) ~[?:?]
        at jdk.internal.util.Preconditions.checkIndex(Preconditions.java:266) ~[?:?]
        at java.util.Objects.checkIndex(Objects.java:359) ~[?:?]
        at org.apache.lucene.index.CodecReader.checkBounds(CodecReader.java:103) ~[lucene-core-9.0.0-snapshot-cc2a31f2be8.jar:9.0.0-snapshot-cc2a31f2be8 cc2a31f2be843935a67c0fdcd3478a65970e791d - mayyasharipova - 2021-11-02 14:36:40]
        at org.apache.lucene.index.CodecReader.document(CodecReader.java:88) ~[lucene-core-9.0.0-snapshot-cc2a31f2be8.jar:9.0.0-snapshot-cc2a31f2be8 cc2a31f2be843935a67c0fdcd3478a65970e791d - mayyasharipova - 2021-11-02 14:36:40]
        at org.apache.lucene.index.FilterLeafReader.document(FilterLeafReader.java:374) ~[lucene-core-9.0.0-snapshot-cc2a31f2be8.jar:9.0.0-snapshot-cc2a31f2be8 cc2a31f2be843935a67c0fdcd3478a65970e791d - mayyasharipova - 2021-11-02 14:36:40]
        at org.apache.lucene.index.FilterLeafReader.document(FilterLeafReader.java:374) ~[lucene-core-9.0.0-snapshot-cc2a31f2be8.jar:9.0.0-snapshot-cc2a31f2be8 cc2a31f2be843935a67c0fdcd3478a65970e791d - mayyasharipova - 2021-11-02 14:36:40]
        at org.elasticsearch.search.internal.FieldUsageTrackingDirectoryReader$FieldUsageTrackingLeafReader.document(FieldUsageTrackingDirectoryReader.java:123) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
        at org.apache.lucene.index.FilterLeafReader.document(FilterLeafReader.java:374) ~[lucene-core-9.0.0-snapshot-cc2a31f2be8.jar:9.0.0-snapshot-cc2a31f2be8 cc2a31f2be843935a67c0fdcd3478a65970e791d - mayyasharipova - 2021-11-02 14:36:40]
        at org.elasticsearch.search.fetch.FetchPhase.loadStoredFields(FetchPhase.java:448) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
        at org.elasticsearch.search.fetch.FetchPhase.prepareNonNestedHitContext(FetchPhase.java:314) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
        at org.elasticsearch.search.fetch.FetchPhase.prepareHitContext(FetchPhase.java:270) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
        at org.elasticsearch.search.fetch.FetchPhase.buildSearchHits(FetchPhase.java:158) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
        at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:90) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
        at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:653) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
        at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:628) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
        at org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:483) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
        at org.elasticsearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:47) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
        at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
        ... 6 more
```
</details>

When running with assertions enabled I run into this check even earlier:
<details>

```
java.lang.AssertionError: null
»       at org.apache.lucene.search.TopScoreDocCollector$SimpleTopScoreDocCollector$1.collect(TopScoreDocCollector.java:75) ~[lucene-core-9.0.0-snapshot-cc2a31f2be8.jar:9.0.0-snapshot-cc2a31f2be8 cc2a31f2be843935a67c0fdcd3478a65970e791d - mayyasharipova - 2021-11-02 14:36:40]
»       at org.apache.lucene.search.Weight$DefaultBulkScorer.scoreRange(Weight.java:274) ~[lucene-core-9.0.0-snapshot-cc2a31f2be8.jar:9.0.0-snapshot-cc2a31f2be8 cc2a31f2be843935a67c0fdcd3478a65970e791d - mayyasharipova - 2021-11-02 14:36:40]
»       at org.apache.lucene.search.Weight$DefaultBulkScorer.score(Weight.java:254) ~[lucene-core-9.0.0-snapshot-cc2a31f2be8.jar:9.0.0-snapshot-cc2a31f2be8 cc2a31f2be843935a67c0fdcd3478a65970e791d - mayyasharipova - 2021-11-02 14:36:40]
»       at org.elasticsearch.search.internal.CancellableBulkScorer.score(CancellableBulkScorer.java:45) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
»       at org.apache.lucene.search.BulkScorer.score(BulkScorer.java:38) ~[lucene-core-9.0.0-snapshot-cc2a31f2be8.jar:9.0.0-snapshot-cc2a31f2be8 cc2a31f2be843935a67c0fdcd3478a65970e791d - mayyasharipova - 2021-11-02 14:36:40]
»       at org.elasticsearch.search.internal.ContextIndexSearcher.searchLeaf(ContextIndexSearcher.java:194) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
»       at org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:167) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
»       at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:541) ~[lucene-core-9.0.0-snapshot-cc2a31f2be8.jar:9.0.0-snapshot-cc2a31f2be8 cc2a31f2be843935a67c0fdcd3478a65970e791d - mayyasharipova - 2021-11-02 14:36:40]
»       at org.elasticsearch.search.query.QueryPhase.searchWithCollector(QueryPhase.java:233) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
»       at org.elasticsearch.search.query.QueryPhase.executeInternal(QueryPhase.java:187) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
»       at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:88) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
»       at org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:458) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
»       at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:621) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
»       at org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:483) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
```
</details>

I also noted that indexing a document with e.g. a [0,0,0] vector into a field with similarity "cosine" is possible and if matched returns "null" scores (or runs into an assertion error if -ea is enabled like in local tests etc...)

Just wanted to mention this, maybe its is useful to protect against this with some better error message etc...

--- a/docs/reference/mapping/types/dense-vector.asciidoc
+++ b/docs/reference/mapping/types/dense-vector.asciidoc
@@ -144,7 +144,9 @@ Computes the cosine similarity. Note that the most efficient way to perform
 cosine similarity is to normalize all vectors to unit length, and instead use
 `dot_product`. You should only use `cosine` if you need to preserve the
 original vectors and cannot normalize them in advance. The document `_score`
-is computed as `(1 + cosine(query, vector)) / 2`.
+is computed as `(1 + cosine(query, vector)) / 2`. The `cosine` similarity does
+not allow vectors with zero magnitude, since cosine is not defined in this
+case.
 ====
 
 NOTE: Although they are conceptually related, the `similarity` parameter is

--- a/x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldMapper.java
+++ b/x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldMapper.java
@@ -310,8 +310,8 @@ public class DenseVectorFieldMapper extends FieldMapper implements PerFieldKnnVe
                 );
             }
 
-            if (similarity == VectorSimilarity.dot_product) {
-                double squaredMagnitude = 0.0;
+            if (similarity == VectorSimilarity.dot_product || similarity == VectorSimilarity.cosine) {
+                float squaredMagnitude = 0.0f;
                 for (float e : queryVector) {
                     squaredMagnitude += e * e;
                 }
@@ -320,28 +320,32 @@ public class DenseVectorFieldMapper extends FieldMapper implements PerFieldKnnVe
             return new KnnVectorQuery(name(), queryVector, numCands);
         }
 
-        private void checkVectorMagnitude(float[] vector, double squaredMagnitude) {
-            if (Math.abs(squaredMagnitude - 1.0f) > 1e-4) {
-                // Include the first five elements of the invalid vector in the error message
-                StringBuilder sb = new StringBuilder(
-                    "The ["
-                        + VectorSimilarity.dot_product.name()
-                        + "] similarity can "
-                        + "only be used with unit-length vectors. Preview of invalid vector: "
+        private void checkVectorMagnitude(float[] vector, float squaredMagnitude) {
+            StringBuilder errorBuilder = null;
+            if (similarity == VectorSimilarity.dot_product && Math.abs(squaredMagnitude - 1.0f) > 1e-4f) {
+                errorBuilder = new StringBuilder(
+                    "The [" + VectorSimilarity.dot_product.name() + "] similarity can " + "only be used with unit-length vectors."
                 );
-                sb.append("[");
+            } else if (similarity == VectorSimilarity.cosine && Math.sqrt(squaredMagnitude) == 0.0f) {
+                errorBuilder = new StringBuilder(
+                    "The [" + VectorSimilarity.cosine.name() + "] similarity does not support vectors with zero magnitude."
+                );
+            }
+
+            if (errorBuilder != null) {
+                // Include the first five elements of the invalid vector in the error message
+                errorBuilder.append(" Preview of invalid vector: [");
                 for (int i = 0; i < Math.min(5, vector.length); i++) {
                     if (i > 0) {
-                        sb.append(", ");
+                        errorBuilder.append(", ");
                     }
-                    sb.append(vector[i]);
+                    errorBuilder.append(vector[i]);
                 }
                 if (vector.length >= 5) {
-                    sb.append(", ...");
+                    errorBuilder.append(", ...");
                 }
-                sb.append("]");
-
-                throw new IllegalArgumentException(sb.toString());
+                errorBuilder.append("]");
+                throw new IllegalArgumentException(errorBuilder.toString());
             }
         }
     }
@@ -399,7 +403,7 @@ public class DenseVectorFieldMapper extends FieldMapper implements PerFieldKnnVe
 
     private Field parseKnnVector(DocumentParserContext context) throws IOException {
         float[] vector = new float[dims];
-        double squaredMagnitude = 0.0;
+        float squaredMagnitude = 0.0f;
         int index = 0;
         for (Token token = context.parser().nextToken(); token != Token.END_ARRAY; token = context.parser().nextToken()) {
             checkDimensionExceeded(index, context);
@@ -410,9 +414,7 @@ public class DenseVectorFieldMapper extends FieldMapper implements PerFieldKnnVe
             squaredMagnitude += value * value;
         }
         checkDimensionMatches(index, context);
-        if (similarity == VectorSimilarity.dot_product) {
-            fieldType().checkVectorMagnitude(vector, squaredMagnitude);
-        }
+        fieldType().checkVectorMagnitude(vector, squaredMagnitude);
         return new KnnVectorField(fieldType().name(), vector, similarity.function);
     }
 

--- a/x-pack/plugin/vectors/src/test/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldMapperTests.java
+++ b/x-pack/plugin/vectors/src/test/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldMapperTests.java
@@ -228,7 +228,7 @@ public class DenseVectorFieldMapperTests extends MapperTestCase {
         assertThat(
             e.getCause().getMessage(),
             containsString(
-                "The [dot_product] similarity can only be used with unit-length vectors. " + "Preview of invalid vector: [-12.1, 2.7, -4.0]"
+                "The [dot_product] similarity can only be used with unit-length vectors. Preview of invalid vector: [-12.1, 2.7, -4.0]"
             )
         );
 
@@ -245,6 +245,23 @@ public class DenseVectorFieldMapperTests extends MapperTestCase {
             containsString(
                 "The [dot_product] similarity can only be used with unit-length vectors. "
                     + "Preview of invalid vector: [-12.1, 2.7, -4.0, 1.05, 10.0, ...]"
+            )
+        );
+    }
+
+    public void testCosineWithZeroVector() throws Exception {
+        DocumentMapper mapper = createDocumentMapper(
+            fieldMapping(
+                b -> b.field("type", "dense_vector").field("dims", 3).field("index", true).field("similarity", VectorSimilarity.cosine)
+            )
+        );
+        float[] vector = { -0.0f, 0.0f, 0.0f };
+        MapperParsingException e = expectThrows(MapperParsingException.class, () -> mapper.parse(source(b -> b.array("field", vector))));
+        assertNotNull(e.getCause());
+        assertThat(
+            e.getCause().getMessage(),
+            containsString(
+                "The [cosine] similarity does not support vectors with zero magnitude. Preview of invalid vector: [-0.0, 0.0, 0.0]"
             )
         );
     }

--- a/x-pack/plugin/vectors/src/test/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldTypeTests.java
+++ b/x-pack/plugin/vectors/src/test/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldTypeTests.java
@@ -85,5 +85,16 @@ public class DenseVectorFieldTypeTests extends FieldTypeTestCase {
         );
         e = expectThrows(IllegalArgumentException.class, () -> dotProductField.createKnnQuery(new float[] { 0.3f, 0.1f, 1.0f }, 10));
         assertThat(e.getMessage(), containsString("The [dot_product] similarity can only be used with unit-length vectors."));
+
+        DenseVectorFieldType cosineField = new DenseVectorFieldType(
+            "f",
+            Version.CURRENT,
+            3,
+            true,
+            VectorSimilarity.cosine,
+            Collections.emptyMap()
+        );
+        e = expectThrows(IllegalArgumentException.class, () -> cosineField.createKnnQuery(new float[] { 0.0f, 0.0f, 0.0f }, 10));
+        assertThat(e.getMessage(), containsString("The [cosine] similarity does not support vectors with zero magnitude."));
     }
 }

